### PR TITLE
Parse fv3 log lines into json

### PR DESCRIPTION
**Pull request by nbren12**
It would be nice to monitor and analzye the fv3 logs more easily to e.g. allow faster turnaround on top level metrics.  Currently, the FV3 logs are not very machine readable and do not integrate well with k8s/gcp logging. This PR processes the logs to output structure json objects that will integrate smoothly with many types of tooling (jq, Google Cloud Logging, etc). One could envision some agent monitoring the stream and uploading top level metrics to wandb.

Some examples:

This is what a fortran output line is transformed to:

    {
      "message": "NOTE from PE     0: MPP_DOMAINS_SET_STACK_SIZE: stack size set to    32768.",
      "logging.googleapis.com/labels": {
        "kind": "fortran",
        "fname": null,
        "model_time": null
      },
      "severity": "DEBUG",
      "model_time": null
    }

A python log line where the message is a json formatted string looks
like this:
```
    {
      "severity": "INFO",
      "module": "statistics",
      "logging.googleapis.com/labels": {
        "kind": "python",
        "fname": null,
        "model_time": "2016-08-10T00:15:00"
      },
      "json": {
        "time": "2016-08-10T00:15:00",
        "column_integrated_dQu": 0,
        "column_integrated_dQv": 0,
        "storage_of_specific_humidity_path_due_to_microphysics": 1.4559948181603119e-05,
        "evaporation": 3.477911480923804e-05,
        "cnvprcp_after_physics": 2.3219503874294952e-05,
        "total_precip_after_physics": 2.78660017111974e-05,
        "storage_of_mass_due_to_fv3_physics": 4.2218462977353356e-05,
        "column_integrated_dQ1_change_non_neg_sphum_constraint": -9.21344804488168,
        "column_integrated_dQ2_change_non_neg_sphum_constraint": 4.112280082814816e-06,
        "net_moistening": -1.0003816217438856e-05,
        "net_heating": 32.11221217099007,
        "water_vapor_path": 25.614354993013656,
        "physics_precip": 3.096222412355266e-05,
        "area": 38505756482.50796,
        "cnvprcp_after_python": 2.3219503874294952e-05,
        "total_precipitation_rate": 4.519084260558481e-05,
        "storage_of_mass_due_to_python": -9.813743709285837e-05
      },
      "model_time": "2016-08-10T00:15:00"
    }
```

I added returns to the examples above for readability....they are
actually emited one json per line, which makes processing with `jq` etc
trivial.

### CODE REVIEW

**Comment by nbren12** [comment]
Here's an example of how google interprets these: 

https://console.cloud.google.com/logs/query;cursorTimestamp=2022-04-06T15:40:52.948838363Z;query=labels.fname%3D%22gs:%2F%2Fvcm-ml-experiments%2Fdefault%2F2021-05-12%2Fhybrid-all-levels%2Ffv3gfs_run%2Fartifacts%2F20160810.000000%2Flogs.txt%22%0Atimestamp%3D%222022-04-06T15:40:52.948023850Z%22%0AinsertId%3D%224jqkp608o9hn8s7v%22;summaryFields=jsonPayload%252Fjson%252Fwater_vapor_path:false:32:beginning:false;timeRange=P7D?project=vcm-ml

File: workflows/prognostic_c48_run/runtime/segmented_run/logs.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+    fname = "/Users/noahb/workspace/ai2cm/fv3net/logs.txt"
+    import fileinput
+
+    with fileinput.input() as f:
+        for line in handle_fv3_log(f, labels={"fname": fileinput.filename()}):
+            print(line)
```
Comment (oliverwm1) [inline workflows/prognostic_c48_run/runtime/segmented_run/logs.py:144]: looks like leftover testing code?

File: workflows/prognostic_c48_run/runtime/segmented_run/logs.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+def parse_line(line):
+    for line_type in LineType:
+        match = REGEX[line_type].match(line)
+        if match:
+            return LogLine(line_type, match.groupdict(), line)
+    raise ValueError(f"Unable to parse line: {line}")
+
+
+class Handler:
+    """Handler handels the parsed lines produces by the log stream
```
Comment (oliverwm1) [inline workflows/prognostic_c48_run/runtime/segmented_run/logs.py:64]:
```suggestion
    """Handler handles the parsed lines produces by the log stream
```

File: workflows/prognostic_c48_run/tests/test_regression.py [hunk 3]
```diff
@@ -727,6 +727,16 @@ def test_fv3run_emulation_zarr_out(completed_rundir, configuration, regtest):
     emu_state_zarr.info(regtest)
 
 
+def test_each_line_of_life_is_json(completed_segment):
```
Comment (oliverwm1) [inline workflows/prognostic_c48_run/tests/test_regression.py:730]:
```suggestion
def test_each_line_of_file_is_json(completed_segment):
```

File: Makefile [hunk 4]
```diff
@@ -26,6 +26,8 @@ build_image_%:
 build_images: $(addprefix build_image_, $(IMAGES))
 push_images: $(addprefix push_image_, $(IMAGES))
 
+build_image_log_handler: docker/log_handler/requirements.txt
```
Comment (oliverwm1) [inline Makefile:29]: i don't see a `docker/log_handler/requirements.txt` in this PR... forget to commit?

**Review by oliverwm1 — COMMENTED** [review]
Looks good, this will enable some cool stuff! Some typos and maybe missing files?

File: Makefile (hunk 4 above)
Comment (oliverwm1) [reply Makefile:29]: or a `requirements.in`...

File: workflows/prognostic_c48_run/runtime/segmented_run/logs.py (hunk 1 above)
Comment (nbren12) [reply workflows/prognostic_c48_run/runtime/segmented_run/logs.py:144]: It's a useful cli for now. It'd be nice to expose this as a public CLI...and extract it and the other runfv3 code from the prognostic run.

Comment (nbren12) [reply workflows/prognostic_c48_run/runtime/segmented_run/logs.py:144]: I added this as a CLI `runfv3 parse-logs`.

File: workflows/prognostic_c48_run/runtime/segmented_run/logs.py (hunk 2 above)
Comment (nbren12) [reply workflows/prognostic_c48_run/runtime/segmented_run/logs.py:64]: I actually meant this guy:

image

File: Makefile (hunk 4 above)
Comment (nbren12) [reply Makefile:29]: sorry. I made a dev image. I'll remove this. Planning to squash so the history won't be to messy.

**Review by oliverwm1 — APPROVED** [review]
LGTM

**Pull request merged**